File: src/signal_model.py
from typing import List, Tuple

from constants import INITIAL_REACTION_WINDOW, ADAPTIVE_WINDOW_CHECK, \
    SUCCESS_THRESHOLD, MIN_REACTION_WINDOW, WINDOW_DECREASE_FACTOR, EXIT_KEY, ENTRY_KEY


class SignalModel:
    """Gère la génération des signaux ENTRY/EXIT et mesure des réactions."""
    def __init__(self) -> None:
        self.reaction_window = INITIAL_REACTION_WINDOW
        self.current_signal: str | None = None
        self.signal_timestamp: float | None = None
        self.results: List[Tuple[float, bool]] = []

    def trigger(self, signal: str, now: float) -> None:
        self.current_signal = signal
        self.signal_timestamp = now

    def record_reaction(self, key: int, now: float) -> bool:
        if self.current_signal is None:
            return False
        rt = now - (self.signal_timestamp or now)
        ok = False
        if ((self.current_signal=="ENTRY" and key==ENTRY_KEY) or
            (self.current_signal=="EXIT" and key==EXIT_KEY)) and rt<=self.reaction_window:
            ok = True
        self.results.append((rt, ok))
        self.current_signal = None
        self._adapt()
        return ok

    def _adapt(self) -> None:
        if len(self.results) < ADAPTIVE_WINDOW_CHECK:
            return
        window = self.results[-ADAPTIVE_WINDOW_CHECK:]
        rate = sum(1 for _, ok in window if ok)/ADAPTIVE_WINDOW_CHECK
        if rate >= SUCCESS_THRESHOLD:
            self.reaction_window = max(
                MIN_REACTION_WINDOW,
                self.reaction_window*WINDOW_DECREASE_FACTOR
            )

File: src/models/SignalModel.ts
// SignalModel.ts - Converted from signal_model.py

import {
  INITIAL_REACTION_WINDOW,
  ADAPTIVE_WINDOW_CHECK,
  SUCCESS_THRESHOLD,
  MIN_REACTION_WINDOW,
  WINDOW_DECREASE_FACTOR,
  ENTRY_KEY,
  EXIT_KEY,
} from "../constants";

export type SignalType = "ENTRY" | "EXIT" | null;

export class SignalModel {
  public reactionWindow: number;
  public currentSignal: SignalType;
  public signalTimestamp: number | null;
  private results: Array<[number, boolean]>;

  constructor() {
    this.reactionWindow = INITIAL_REACTION_WINDOW;
    this.currentSignal = null;
    this.signalTimestamp = null;
    this.results = [];
  }

  trigger(signal: "ENTRY" | "EXIT", now: number): void {
    this.currentSignal = signal;
    this.signalTimestamp = now;
  }

  recordReaction(key: string, now: number): boolean {
    if (this.currentSignal === null || this.signalTimestamp === null) {
      return false;
    }

    const rt = now - this.signalTimestamp;
    let ok = false;

    if (
      ((this.currentSignal === "ENTRY" && key === ENTRY_KEY) ||
        (this.currentSignal === "EXIT" && key === EXIT_KEY)) &&
      rt <= this.reactionWindow
    ) {
      ok = true;
    }

    this.results.push([rt, ok]);
    this.currentSignal = null;
    this.adapt();
    return ok;
  }

  private adapt(): void {
    if (this.results.length < ADAPTIVE_WINDOW_CHECK) {
      return;
    }

    const window = this.results.slice(-ADAPTIVE_WINDOW_CHECK);
    const successCount = window.filter(([_, ok]) => ok).length;
    const rate = successCount / ADAPTIVE_WINDOW_CHECK;

    if (rate >= SUCCESS_THRESHOLD) {
      this.reactionWindow = Math.max(
        MIN_REACTION_WINDOW,
        this.reactionWindow * WINDOW_DECREASE_FACTOR
      );
    }
  }

  getResults(): Array<[number, boolean]> {
    return [...this.results];
  }
}
